Cargo.toml

```
@@ -1,21 +1,18 @@
 [package]
 name = "multibase"
 version = "0.7.0"
-description = "multibase rust"
-homepage = "https://github.com/multiformats/rust-multibase"
 authors = ["Friedel Ziegelmayer <dignifiedquire@gmail.com>"]
 edition = "2018"
-
-keywords = ["ipld", "ipfs", "multihash", "multibase", "cid"]
-
 license = "MIT"
-
 readme = "README.md"
+description = "multibase in rust"
+homepage = "https://github.com/multiformats/rust-multibase"
+keywords = ["ipld", "ipfs", "multihash", "multibase", "cid"]
 
 [dependencies]
 base-x = "0.2"
-data-encoding = "2.1.2"
-data-encoding-macro = "0.1.7"
+data-encoding = "2.1"
+lazy_static = "1.4"
```

Is there a reason to not do

```suggestion
base-x = "^0.2"
data-encoding = "^2.1"
lazy_static = "^1.4"
```

---

Because it's the same

---

`^1.4` isn't the same as `1.4` -- `^1.4` will allow automatically updating to use `1.9999999`, while `1.4` won't.

---

See [doc](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio) for details.

> The string "0.1.12" is a semver version requirement. Since this string does not have any operators in it, it is interpreted the same way as if we had specified "^0.1.12", which is called a caret requirement.

---

That's true for pre-`1.0` projects, but isn't true for post-1.0 projects. From the same link:

> If instead we had specified the version string as ^1.0, cargo should update to 1.1 if it is the latest 1.y release, but not 2.0.

---

"0.1.12" is just a example...
What the document actually wants to express is that `"x.y.z"` is interpreted  the same way as if we had specified `"^x.y.z"`

---

I quoted the document you linked where it explicitly explains how caret requirements work. It *does not* say that caret requirements are a no-op -- what use would the feature have if it did nothing? If you don't believe me, try build a crate that has dependencies like `lazy_static = "^1.0"` -- it will use `lazy_static 1.4.0`:

```
% grep lazy_static Cargo.toml
lazy_static = "^1.0"
% cargo build
    Updating crates.io index
   Compiling lazy_static v1.4.0
   Compiling test1 v0.1.0 (/home/cyphar/tmp/test1)
    Finished dev [unoptimized + debuginfo] target(s) in 4.86s
```

But okay, I was only making a suggestion. If you don't want to do it, that's fine.

---

Have you tried to build a crate with lazy_static=1.0? It will use lazy_static v1.4 too.

---

Oh... You're quite right. Oops.